### PR TITLE
Unicode filenames

### LINKED ISSUE #4507: filenames with unicode characters should keep the original file name when saving to disk
### reproduction
* Upload a resource which has a file name with unicode characters
* View the resource page and inspect the download link

### expected
* Download link should contain the original filename with the unicode characters

### actual
* Download link truncates the unicode characters

## example
* https://demo.ckan.org/dataset/test-hebrew/resource/240cabfe-0057-4aea-9ce0-dbfb708bc484
* original file name: `נוכחות בישיבות ועדה 1-5-2017 עד 30-7-2017.xlsx`
* download url: `https://demo.ckan.org/dataset/0d9acb2e-638a-4d73-8511-2ff52a7f654a/resource/240cabfe-0057-4aea-9ce0-dbfb708bc484/download/---1-5-2017--30-7-2017.xlsx`

### PR DESCRIPTION
**Pull request by frafra**
Fixes #4507 

### Proposed fixes:
Do not remove/replace Unicode characters with 7-bit ASCII and allow Unicode letters and numbers.

### Features:

- [x] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

### CODE REVIEW

**Comment by wardi** [comment]
We should replace all non-"letter" category characters, i.e.
```python
not unicodedata.category(char).startswith('L')
```

**Comment by frafra** [comment]
Unicode letters are ok, Unicode non-letters are not. Got it :-)

**Comment by amercader** [comment]
@frafra Many thanks for your patch. As @wardi mentioned we still need to strip non-letter characters so IIRC your changes in e8f0631a1 are not needed (ie these tests should still pass unchanged).

Perhaps is useful to update the tests in test_munge.py with some examples like the ones @OriHoch mentions in the issue and work from there.

**Comment by wardi** [comment]
Hmm. numbers should be allowed too. Better to list the specific categories we'll allow
http://www.fileformat.info/info/unicode/category/index.htm

@amercader what do you think of allowing:

Code | Description
-- | --
[Ll] | Letter, Lowercase
[Lm] | Letter, Modifier
[Lo] | Letter, Other
[Lt] | Letter, Titlecase
[Lu] | Letter, Uppercase
[Nd] | Number, Decimal Digit

**Comment by wardi** [comment]
https://pypi.org/project/regex/ looks useful, it can match unicode categories without looping over every character

**Comment by amercader** [comment]
> @amercader what do you think of allowing:

Just for completeness should we maybe also include the following?

Code | Description
-- | --
[Nl] | Number, Letter
[No] | Number, Other

**Comment by frafra** [comment]
If we include every number and letter we could just use this regex (plus few additional characters like `-_.` that are specified inside the functions):

```python
regex.sub(r'[^\p{NUMBER}&&\p{LETTER}]', '', text)
```

I would also have to add a new requirement for `regex`. Let me know, so I can do a new commit.

**Comment by frafra** [comment]
Ok, it seems pretty clean to me. Commits can be squashed when merged and/or more tests could be added.

File: ckan/lib/munge.py [hunk 1]
```diff
@@ -134,9 +77,8 @@ def munge_filename_legacy(filename):
     Upload.update_data_dict() because if this routine changes then group images
     uploaded previous to the change may not be viewable.
     '''
-    filename = substitute_ascii_equivalents(filename)
     filename = filename.strip()
-    filename = re.sub(r'[^a-zA-Z0-9.\- ]', '', filename).replace(' ', '-')
```
Comment (wardi) [inline ckan/lib/munge.py]: the name and docstring of this function suggest we shouldn't change anything here. If we can remove this it would be great as a separate PR

Comment (frafra) [reply ckan/lib/munge.py]: @wardi Do you suggest to keep using `substitute_ascii_equivalents` in this function for this PR?

Comment (frafra) [reply ckan/lib/munge.py]: Right.  Removed.

**Comment by frafra** [comment]
All test are go.
@OriHoch would you like to try this branch and see if it fixes the bug you reported, before merging to master?

**Comment by OriHoch** [comment]
thanks, looks like there is something wrong with the encoding, this is what I'm getting - 


image

This is the file (remove .txt extension):

[נתוני נוכחות במליאה לפי נתוני הצבעה - כנסת 20 - מעודכנים ל18 ליולי 2018.ods.txt](https://github.com/ckan/ckan/files/2535112/-.20.-.18.2018.ods.txt)

(looks like github are munging it too..)

**Comment by frafra** [comment]
Thank you @OriHoch. I think this is due to CKAN default `schema.xml`, which uses [ASCIIFoldingFilter](https://lucene.apache.org/solr/guide/6_6/filter-descriptions.html#FilterDescriptions-ASCIIFoldingFilter) for text strings. Could you please test again?

There is just one legacy test failing, so if it works for you, I am going to find a solution. Thanks!

**Comment by OriHoch** [comment]
thanks, will test later today

**Comment by OriHoch** [comment]
The bug persists, 

this is how the URL is stored in the DB - 
image

and this is in SOLR - 
image

**Comment by frafra** [comment]
Does anyone has a clue about it? It seems something related more to Solr configuration that to CKAN.

**Comment by kmbn** [comment]
@frafra Are you still working on this PR? In addition to the issue @OriHoch mentioned, the tests fail on CircleCI.

**Comment by frafra** [comment]
@kmbn The patch is halfway trough, but as I wrote in the last comment I would need help to fix it, because it looks like Solr is involved.

**Comment by pwalsh** [comment]
@frafra @amercader @wardi 

Instead of custom munging, would it be possible (and easier, and well tested) to use the [Mozilla slugify library](https://github.com/mozilla/unicode-slugify)? I have used this extensively on projects that require unicode URLs.

**Comment by tino097** [comment]
This PR is old and some of the changes could not apply or have been fixed within other PR's
